### PR TITLE
Filter out migrating endpoints in NEG controller

### PR DESCRIPTION
**Pull request by gauravkghildiyal**
/assign @swetharepakula

### CODE REVIEW

File: pkg/neg/syncers/utils_test.go [hunk 1]
```diff
@@ -300,6 +300,236 @@ func TestNetworkEndpointCalculateDifference(t *testing.T) {
...
+
+	testCases := []struct {
+		desc                              string
+		addEndpoints                      map[string]negtypes.NetworkEndpointSet
+		removeEndpoints                   map[string]negtypes.NetworkEndpointSet
+		wantMigrationEndpointsInAddSet    map[string]negtypes.NetworkEndpointSet
+		wantMigrationEndpointsInRemoveSet map[string]negtypes.NetworkEndpointSet
+	}{
+		{
+			desc: "normal",
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils_test.go:314]: expand the description, what does normal mean?

File: pkg/neg/syncers/utils_test.go [hunk 2]
```diff
@@ -300,6 +300,236 @@ func TestNetworkEndpointCalculateDifference(t *testing.T) {
...
+		endpoint    negtypes.NetworkEndpoint
+		inputSource map[string]negtypes.NetworkEndpointSet
+		inputDest   map[string]negtypes.NetworkEndpointSet
+		wantSource  map[string]negtypes.NetworkEndpointSet
+		wantDest    map[string]negtypes.NetworkEndpointSet
+		zone        string
+		wantSuccess bool
+	}{
+		{
+			desc:     "normal",
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils_test.go:417]: expand on what normal means

File: pkg/neg/syncers/utils_test.go [hunk 3]
```diff
@@ -300,6 +300,236 @@ func TestNetworkEndpointCalculateDifference(t *testing.T) {
...
+					{IP: "a", IPv6: "A"},
+				}...),
+				"zone2": negtypes.NewNetworkEndpointSet([]negtypes.NetworkEndpoint{
+					{IP: "c", IPv6: "C"},
+				}...),
+			},
+			zone:        "zone1",
+			wantSuccess: true,
+		},
+		{
```
Comment (swetharepakula) [inline pkg/neg/syncers/utils_test.go:499]: one more test case for when dest is nil

**Review by swetharepakula — COMMENTED** [review]
only nits. this is almost ready to be merged

File: pkg/neg/syncers/utils_test.go (hunk 3 above)
Comment (gauravkghildiyal) [reply pkg/neg/syncers/utils_test.go:499]: Done!

File: pkg/neg/syncers/utils.go [hunk 4]
```diff
@@ -99,6 +99,75 @@ func calculateNetworkEndpointDifference(targetMap, currentMap map[string]negtype
...
+
+			// At this point, we know that `endpoint` is a dual-stack endpoint. An
+			// endpoint can be identified as migrating if the IPs from the dual-stack
+			// endpoint exist as individual single-stack endpoint inside
+			// `addEndpoints` or `removeEndpoints`.
+
+			ipv4Only := endpoint
+			ipv4Only.IPv6 = ""
+			ipv6Only := endpoint
+			ipv6Only.IP = ""
```
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/syncers/utils.go]: Done!

File: pkg/neg/syncers/utils.go [hunk 5]
```diff
@@ -99,6 +99,75 @@ func calculateNetworkEndpointDifference(targetMap, currentMap map[string]negtype
...
+
+			isMigrating := false
+			isMigrating = isMigrating || moveEndpoint(ipv4Only, addEndpoints, migrationEndpointsInAddSet, zone)
+			isMigrating = isMigrating || moveEndpoint(ipv6Only, addEndpoints, migrationEndpointsInAddSet, zone)
+			isMigrating = isMigrating || moveEndpoint(ipv4Only, removeEndpoints, migrationEndpointsInRemoveSet, zone)
+			isMigrating = isMigrating || moveEndpoint(ipv6Only, removeEndpoints, migrationEndpointsInRemoveSet, zone)
```
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/syncers/utils.go]: Done!

File: pkg/neg/syncers/utils.go [hunk 6]
```diff
@@ -99,6 +99,75 @@ func calculateNetworkEndpointDifference(targetMap, currentMap map[string]negtype
...
+				moveEndpoint(endpoint, removeEndpoints, migrationEndpointsInRemoveSet, zone)
+			}
+		}
+	}
+
+	return migrationEndpointsInAddSet, migrationEndpointsInRemoveSet
+}
+
+// moveEndpoint deletes endpoint `e` from `source[zone]` and adds it to
+// `dest[zone]`. If the move was successful, `true` is returned.
```
[the remark this answers is not in the document]
Comment (gauravkghildiyal) [reply pkg/neg/syncers/utils.go]: Incorporated.

File: pkg/neg/syncers/utils_test.go (hunk 2 above)
Comment (gauravkghildiyal) [reply pkg/neg/syncers/utils_test.go]: Changed the field name to "name" instead of "desc" to meet expectation.

**Review by gauravkghildiyal — COMMENTED** [review]
Thanks for the review!

**Comment by gauravkghildiyal** [comment]
Ready.

**Review by swetharepakula — COMMENTED** [review]
/lgtm
/approve
